### ISSUE TITLE
VW daemon protocol errors

### ISSUE THREAD

**Issue by marko-asplund** [JohnLangford/vowpal_wabbit#642]
I'm experiencing protocol errors with the daemon protocol when trying to upgrade VW to a version built from source code at commit 882f34f8cc.

I fit an LDA model like this:
```
$VW -d $HOME/vowpal_wabbit/test/train-sets/wiki256.dat --lda 100 --lda_alpha 0.1 --lda_rho 0.1 --lda_D 10 --minibatch 10 --power_t 0.5 --initial_t 1 -b 8 --cache_file temp/vw.cache --passes 2 --readable_model data/topics.dat --holdout_off -f data/predictor.vw
```

And then run VW in daemon mode to get predictions:
```
$VW -i data/predictor.vw -t --daemon --active --learning_rate 0.0 --lda_epsilon 100000
```

here's output from a sample session for requesting predictions:

server-side:
```
marko@foo:~/vw/daemon1$ $VW -i data/predictor.vw -t --daemon --active --learning_rate 0.0 --lda_epsilon 100000
only testing
Num weight bits = 8
learning rate = 0
initial_t = 0
power_t = 0.5
using no cache
calling accept
// request #1
reading data from port 26542
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
9.686386 9.686386            1            1.0  unknown   0.0000        2
write error: Bad file descriptor
4.843193 0.000000            2            2.0  unknown   0.0000        2
// request #2
write error: Bad file descriptor
4.843193 4.843193            4            4.0  unknown   0.0000        2
// request #3
// request #4
4.843193 4.843193            8            8.0  unknown   0.0000        2
```

client-side:
```
marko@foo:~$ echo '| 5354:1 59:2' | nc localhost 26542
0.142728 0.157917 0.112893 0.105314 0.172945 0.108496 0.134989 0.122645 0.123844 0.108044 0.116526 0.109928 0.164178 0.112116 0.126857 0.140717 0.129266 0.199520 0.106527 0.107983 0.108106 0.112010 0.151527 0.127467 0.179960 0.115240 0.195877 0.113511 0.105400 0.121025 0.163396 0.110505 0.105549 0.110881 0.108623 0.108845 0.127518 0.216021 0.156641 0.116267 0.114576 0.109681 0.202836 0.114521 0.107254 0.132266 0.113445 0.103883 0.126514 0.193035 0.104652 0.121841 0.103883 0.125267 0.206758 0.115217 0.223736 0.104459 0.108299 0.125356 0.109275 0.221981 0.118299 0.119880 0.125523 0.127159 0.140248 0.130185 0.113630 0.139731 0.131302 0.113976 0.138060 0.107683 0.141779 0.110657 0.120478 0.110673 0.120482 0.105087 0.108752 0.115903 0.156564 0.123090 0.141717 0.146593 0.111871 0.109617 0.146141 0.142696 0.120598 0.110337 0.206528 0.112337 0.106998 0.104635 0.128801 0.108227 0.103785 0.133443
marko@foo:~$ echo '| 5354:1 59:2' | nc localhost 26542
marko@foo:~$ echo '| 5354:1 59:2' | nc localhost 26542
marko@foo:~$ echo '| 5354:1 59:2' | nc localhost 26542
0.142728 0.157917 0.112893 0.105314 0.172945 0.108496 0.134989 0.122645 0.123844 0.108044 0.116526 0.109928 0.164178 0.112116 0.126857 0.140717 0.129266 0.199520 0.106527 0.107983 0.108106 0.112010 0.151527 0.127467 0.179960 0.115240 0.195877 0.113511 0.105400 0.121025 0.163396 0.110505 0.105549 0.110881 0.108623 0.108845 0.127518 0.216021 0.156641 0.116267 0.114576 0.109681 0.202836 0.114521 0.107254 0.132266 0.113445 0.103883 0.126514 0.193035 0.104652 0.121841 0.103883 0.125267 0.206758 0.115217 0.223736 0.104459 0.108299 0.125356 0.109275 0.221981 0.118299 0.119880 0.125523 0.127159 0.140248 0.130185 0.113630 0.139731 0.131302 0.113976 0.138060 0.107683 0.141779 0.110657 0.120478 0.110673 0.120482 0.105087 0.108752 0.115903 0.156564 0.123090 0.141717 0.146593 0.111871 0.109617 0.146141 0.142696 0.120598 0.110337 0.206528 0.112337 0.106998 0.104635 0.128801 0.108227 0.103785 0.133443
0.000000  1.000000
// client side "hangs" at this point
^C
```

With the old version built from source 4d7b712e5f this works.

Is this a bug or has the daemon protocol changed somehow?

I'm using Ubuntu 13.10 (x86_64).

**Comment by sharatsc** [comment]
Looks like a bug. I will take a look

**Comment by sharatsc** [comment]
Hi, 
a) why are you specifying --active?
b) Without --active, I don't seem to have any problems.
c) I tried undoing changes to lda_core.cc
https://github.com/sharatsc/vowpal_wabbit/commit/688f277799f06211152e7a985cddec3be6957b0d 
I think this may be unrelated to changes in lda_core.cc .

Can you confirm?

**Comment by marko-asplund** [comment]
Hi,
thanks @sharatsc.

a) started wondering about that, too, now that you mentioned it. Is the VW LDA implementation a purely unsupervised one or does it also support semi-supervised mode via active learning? Also, I guess that `--learning_rate 0.0` would imply that the VW gradient descent algorithm wouldn't actually change the parameter values between iterations (since alpha = 0 would cancel out the derivative term in each update step), and thus no learning would take place, right?
(The command line parameters have been selected a while ago by another person, so I'll have to do a bit of digging to find out the rationale behind using them.)

b) if i drop the `--active` flag I don't get the protocol errors anymore.
In this case the text "calling accept" is output on the console 10 times whereas previously only once.
Also, when I terminate the foreground process using CTRL-C 11 VW processes are left running in the background. It seems strange that this parameter would have such an effect on the daemon.

c) I cloned your repository, checked out commit 688f277799 and re-ran my test with that version, but I still get the same behaviour as before.

**Comment by sharatsc** [comment]
Sent from my iPhone

> On May 22, 2015, at 6:38 AM, marko asplund <notifications@github.com> wrote:
> 
> Hi,
> thanks @sharatsc.
> 
> a) started wondering about that, too, now that you mentioned it. Is the VW LDA implementation a purely unsupervised one or does it also support semi-supervised mode via active learning? Also, I guess that --learning_rate 0.0 would imply that the VW gradient descent algorithm wouldn't actually change the parameter values between iterations (since alpha = 0 would cancel out the derivative term in each update step), and thus no learning would take place, right?
> (The command line parameters have been selected a while ago by another person, so I'll have to do a bit of digging to find out the rationale behind using them.)
> 
--active option is not used anywhere in the LDA code. -t (test only) has the same effect of learning rate =0
> b) if i drop the --active flag I don't get the protocol errors anymore.
> In this case the text "calling accept" is output on the console 10 times whereas previously only once.
> Also, when I terminate the foreground process using CTRL-C 11 VW processes are left running in the background. It seems strange that this parameter would have such an effect on the daemon.
> 
By default num_children=10. There Tara 10 processes answering requests from the same model. You can set it to 1.
> c) I cloned your repository, checked out commit 688f277 and re-ran my test with that version, but I still get the same behaviour as before.
> 
Is this without --active?

**Comment by marko-asplund** [comment]
>> c) I cloned your repository, checked out commit 688f277 and re-ran my test with that version, but I still get the same behaviour as before.

>Is this without --active?

When I specify the `--active` flag with version 688f277799 I still get the "Bad file descriptor" (as well as the other errors described above). With the `--active` flag version 688f277 seems to spawn only one process and it's a foreground process (attached to the terminal), but without the flag 11 background processes get spawned and the group leader detaches from the terminal.
The only difference between the versions seems to be that with version 688f277799 I only get one "calling accept" message on the console (without `--active`), whereas with 882f34f8cc I get this 10 times.

It seems strange that the use of a flag that's not handled by the LDA implementation leads to protocol errors. Another way to look at this would be that if active learning is inconsistent with the LDA variant that VW implements maybe a warning or error message should be output when these features are selected in conjunction with each other? Especially if using these together leads to errors?

**Comment by sharatsc** [comment]
The default for num_children is 10  (from https://github.com/JohnLangford/vowpal_wabbit/wiki/Command-line-arguments)

Looking at usages of --active, it is not meant to be used along with --lda C:\Users\shchikke\Documents\vowpal_wabbit-7.4\vowpal_wabbit-7.4\vowpalwabbit/global_data.cc:  num_children = 10;

I'm not very well versed in how --active operates, but there are pieces of code where daemon is mean to be enabled. From this it looks like daemon should be disabled when --active is true.
  if (vm.count("daemon") || vm.count("pid_file") || (vm.count("port") && !all.active) ) { 
    all.daemon = true; 

In either case, can we agree that this is unrelated to --lda? The protocol errors are due to interaction of --daemon and --active flags.

**Comment by sharatsc** [comment]
We should keep the issue open. I'll investigate the --daemon and --active interaction unless somebody else wants to own it.

**Comment by arielf** [comment]
I wouldn't make `--active` and `--daemon` exclusive. Isn't this a fix that is worse than the problem?
Daemonizing seems to have some issues (unrelated to LDA): it needs to detach from the terminal (so ^C doesn't affect it), and if `isatty(stdout)` it should redirect them to `/dev/null` (same for `stdin` and `stdout`).  All this should be taken care of by the library `daemon (3)` function (`man 3 daemon`, unsupported on Windows, and perhaps soon on OS-X). So, when `vw` calls the `daemon(nochdir, noclose)` function this should all be taken care of. I think part of the problem may be that by then all the children are already forked so `daemon()` doesn't affect the children. There's a separate issue with signal catching: the parent process calls `sigaction(SIGTERM, &sa, nullptr)` in `parser.cc` (ignores the `TERM` signal) so result is that `killall vw` kills the children only, and leaves the parent alive.

**Comment by JohnLangford** [comment]
--active is not valid with --lda because --active relies on a supervised 
learning algorithm. --active supports active learning, for which you 
typically want a much more controlled interaction process than daemon 
gives.  I've added a warning+thrown exception.

Ariel, is there a better way to handle daemon termination?

-John

**Comment by JohnLangford** [comment]
Closing this for now, but I am interested in any better ways to terminate a daemon.

**Comment by arielf** [comment]
The issue I experienced with daemon termination was that `pkill vw` would kill the children and leave one (the parent?) process alive.  Only when I switched to `pkill -9 vw` all processes would be killed.  This is why `test/daemon-test.sh` in the tree uses `$PKILL -9`.

A separate issue was that standard FDs (stdin, stdout, stderr) were not closed as is the standard practice in daemons.  This may actually be seen as a feature because it allowed me (in a different context at work) to redirect them myself to my log of choice, like this:

```
$DaemonCmd="vw -i modelfile -t --daemon --num_children 20 --quiet --port 20001";
$DaemonCmd </dev/null 2>&1 | logger -i -t vw &
```

So that errors (e.g. malformed example cases) would be logged into syslog with the `vw` tag.
 
Is this helpful?

**Comment by JohnLangford** [comment]
Why not kill the parent based on pid so that it does a clean shutdown?

We could close stdin/stdout/stderr, but it's kind of handy to see 
diagnostic output.  What's the utility?

-John

**Comment by arielf** [comment]
1) Often you don't have the parent PID (e.g. in a shell script) so a `killall vw` or `pkill vw` are the most natural way to terminate.  When the parent gets SIGTERM should it terminate after killing the children?

2) Indeed. stdout and stderr are useful, so please leave as is (caller can redirect).  But shouldn't stdin at least be closed and disscociated from the terminal when the reading switches to the TCP socket?

**Comment by marko-asplund** [comment]
@arielf I use `--pid_file` to get parent PID written to a file and send the `TERM` signal to the parent process to shut down VW. This seems to work ok, but as you noted, using the `--daemon` flag doesn't work quite daemonize VW in the same way as with many other daemon programs esp. wrt. STDERR and STDOUT (these could be redirected to a separate log file when using VW in daemon mode).

**Comment by JohnLangford** [comment]
I added an fclose(stdin) to daemon mode to address 2.  I'm not sure what 
can or should be done about (1). Supporting a clean kill semantics via 
PID seems too desirable to lose, so the question is: Can we also support 
a dirty kill?

-John
